### PR TITLE
Split relay auth config and handle multiple addresses

### PR DESCRIPTION
**Pull request by lixmal**
### Checklist
- [x] Is it a bug fix
- [ ] Is a typo/documentation fix
- [x] Is a feature enhancement
- [x] It is a refactor
- [ ] Created tests that fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: relay/client/manager.go [hunk 1]
```diff
@@ -86,29 +88,66 @@ func (m *Manager) Serve() error {
...
-	})
-	m.startCleanupLoop()
+	// Wait for the first successful connection or all attempts to fail
+	select {
+	case client := <-successChan:
+		log.Infof("Successfully connected to relay server: %s", client.connectionURL)
+
+		m.relayClient = client
+
+		m.reconnectGuard = NewGuard(m.ctx, m.relayClient)
```
Comment (mlsmaycon) [inline relay/client/manager.go:115]: What do you think about refactoring NewGuard to use the URL list as a fallback option?

File: relay/auth/hmac/token.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ func unmarshalToken(payload []byte) (Token, error) {
 	return creds, err
 }
 
-// TimedHMAC generates token with TTL and using pre-shared secret known to TURN server
+// TimedHMAC generates token with TTL and using pre-shared secret known to the TURN/relay server
```
Comment (mlsmaycon) [inline relay/auth/hmac/token.go]:
```suggestion
// TimedHMAC generates a token with TTL and uses a pre-shared secret known to the relay server
```
With the new config, the key doesn't necessarily needs to be known by TURN

File: relay/auth/hmac/token_test.go [hunk 3]
```diff
@@ -12,7 +14,7 @@ func TestGenerateCredentials(t *testing.T) {
 	timeToLive := 1 * time.Hour
 	v := NewTimedHMAC(secret, timeToLive)
 
-	creds, err := v.GenerateToken()
+	creds, err := v.GenerateToken(sha1.New)
```
Comment (mlsmaycon) [inline relay/auth/hmac/token_test.go:17]: shouldn't we test with `shar256.New`?

File: relay/auth/hmac/token_test.go [hunk 4]
```diff
@@ -97,7 +99,7 @@ func TestInvalidPayload(t *testing.T) {
 		Signature: creds.Signature,
 	}
 
-	if err = v.Validate(*invalidPayloadCreds); err == nil {
+	if err = v.Validate(sha1.New, *invalidPayloadCreds); err == nil {
```
Comment (mlsmaycon) [inline relay/auth/hmac/token_test.go:102]: mismatch alg

File: relay/client/manager.go [hunk 5]
```diff
@@ -86,29 +88,66 @@ func (m *Manager) Serve() error {
...
+		}(url)
 	}
 
-	m.reconnectGuard = NewGuard(m.ctx, m.relayClient)
-	m.relayClient.SetOnDisconnectListener(func() {
-		m.onServerDisconnected(m.serverURL)
-	})
-	m.startCleanupLoop()
+	// Wait for the first successful connection or all attempts to fail
+	select {
```
Comment (pappz) [inline relay/client/manager.go:109]: What is the goal of this logic?
How will look the server, and load balancer infrastructure? This parallel connection attempt approach will not be ended on the same server with two different URLs? The original idea was that every peer tries to connect to the same URL and the server-side load balancer manages the server instance assigmenet.

Comment (pappz) [inline relay/client/manager.go:109]: If we still want to add additional transport methods (WS, UDP, custom TCP based things) to support the wide range of firewalls with the fastest way of communications, then this parallel connection attempt could cause overload on the server load balancer. I.e. the MGM server distributes 2 server addresses and we will support WS, UDP, custom TCP connections then every peer will send 6 connection attempts.

File: relay/auth/hmac/token_test.go (hunk 3 above)
Comment (lixmal) [reply relay/auth/hmac/token_test.go:17]: Fixed

File: relay/server/peer.go [hunk 6]
```diff
@@ -162,22 +172,30 @@ func (p *Peer) healthcheck(ctx context.Context, hc *healthcheck.Sender) {
 }
 
 func (p *Peer) handleTransportMsg(msg []byte) {
-	peerID, err := messages.UnmarshalTransportID(msg)
+	version, peerID, err := messages.UnmarshalTransportID(msg)
```
Comment (pappz) [inline relay/server/peer.go:176]: Already checked the version in the read loop based on the same information. I think here enough to handle the peerID and payload.

Comment (lixmal) [reply relay/server/peer.go:176]: Fixed

File: relay/auth/hmac/token_test.go (hunk 4 above)
Comment (lixmal) [reply relay/auth/hmac/token_test.go:102]: Fixed

File: client/testdata/management.json [hunk 7]
```diff
@@ -20,6 +20,11 @@
     "Secret": "c29tZV9wYXNzd29yZA==",
     "TimeBasedCredentials": true
   },
+  "Relay": {
+    "Address": "localhost:0",
```
Comment (pappz) [inline client/testdata/management.json]: This should be a list

File: relay/messages/message.go [hunk 8]
```diff
@@ -55,37 +59,53 @@ type HelloResponse struct {
...
-	case MsgTypeHealthCheck:
+	case
+		MsgTypeHello,
+		MsgTypeTransport,
+		MsgTypeClose,
+		MsgTypeHealthCheck:
 		return msgType, nil
 	default:
-		return 0, fmt.Errorf("invalid msg type, len: %d", len(msg))
+		return 0, fmt.Errorf("invalid msg type %d, len: %d", msgType, len(msg))
```
Comment (pappz) [inline relay/messages/message.go]: The len(msg) is invalid because this function get the buf[1:n]

File: relay/messages/message.go [hunk 9]
```diff
@@ -55,37 +59,53 @@ type HelloResponse struct {
 	InstanceAddress string
 }
 
-// DetermineClientMsgType determines the message type from the first byte of the message
-func DetermineClientMsgType(msg []byte) (MsgType, error) {
+// ValidateVersion checks if the given version is supported by the protocol
+func ValidateVersion(msg []byte) (int, error) {
```
Comment (pappz) [inline relay/messages/message.go:64]: If I look right we never use the returned int.

Comment (lixmal) [reply relay/messages/message.go:64]: Correct, but might use it if we need to branch based on version

File: relay/messages/message.go (hunk 8 above)
Comment (lixmal) [reply relay/messages/message.go]: fixed

File: client/testdata/management.json (hunk 7 above)
Comment (lixmal) [reply client/testdata/management.json]: fixed

File: relay/messages/message.go [hunk 10]
```diff
@@ -55,37 +59,53 @@ type HelloResponse struct {
...
-	case MsgTypeHealthCheck:
+	case
+		MsgTypeHelloResponse,
+		MsgTypeTransport,
+		MsgTypeClose,
+		MsgTypeHealthCheck:
 		return msgType, nil
 	default:
-		return 0, fmt.Errorf("invalid msg type (len: %d)", len(msg))
+		return 0, fmt.Errorf("invalid msg type %d, len: %d", msgType, len(msg))
```
Comment (pappz) [inline relay/messages/message.go]: Message len is invalid because we skip the first byte of the msg.

**Review by pappz — APPROVED** [review]

File: relay/client/manager.go (hunk 1 above)
Comment (lixmal) [reply relay/client/manager.go:115]: I think we can make this improvement soon after release

**Pull request merged**